spring-kafka/src/test/java/org/springframework/kafka/streams/RecoveringDeserializationExceptionHandlerTests.java

```diff
@@ -0,0 +1,280 @@
+				new IllegalStateException())).isEqualTo(DeserializationHandlerResponse.FAIL);
+	}
+
+	@Test
+	void withNoRecoverer() {
+		RecoveringDeserializationExceptionHandler handler = new RecoveringDeserializationExceptionHandler();
+		assertThat(handler.handle(null, new ConsumerRecord<byte[], byte[]>("foo", 0, 0, null, null),
+				new IllegalArgumentException())).isEqualTo(DeserializationHandlerResponse.FAIL);
+	}
+
+	@Test
+	void integration() throws Exception {
+		this.kafkaTemplate.send("recoverer1", "foo".getBytes(), "bar".getBytes());
+		ConsumerRecord<byte[], byte[]> record = this.resultFuture.get(10, TimeUnit.SECONDS);
+		assertThat(record).isNotNull();
+		assertThat(record.key()).isEqualTo("foo".getBytes());
+		assertThat(record.value()).isEqualTo("bar".getBytes());
+		assertThat(record.headers().lastHeader(KafkaHeaders.DLT_EXCEPTION_STACKTRACE)).isNotNull();
+	}
+
+	@Configuration
+	@EnableKafka
+	@EnableKafkaStreams
+	public static class KafkaStreamsConfig {
+
+		@Value("${" + EmbeddedKafkaBroker.SPRING_EMBEDDED_KAFKA_BROKERS + "}")
+		private String brokerAddresses;
+
+		@Value("${streaming.topic.two}")
+		private String streamingTopic2;
+
+		@Bean
+		public ProducerFactory<byte[], byte[]> producerFactory() {
+			return new DefaultKafkaProducerFactory<>(producerConfigs());
+		}
+
+		@Bean
+		public Map<String, Object> producerConfigs() {
+			Map<String, Object> senderProps = KafkaTestUtils.senderProps(this.brokerAddresses);
+			senderProps.put(ProducerConfig.KEY_SERIALIZER_CLASS_CONFIG, ByteArraySerializer.class);
+			senderProps.put(ProducerConfig.VALUE_SERIALIZER_CLASS_CONFIG, ByteArraySerializer.class);
+			return senderProps;
+		}
+
+		@Bean
+		public KafkaTemplate<byte[], byte[]> template() {
+			KafkaTemplate<byte[], byte[]> kafkaTemplate = new KafkaTemplate<>(producerFactory(), true);
+			kafkaTemplate.setDefaultTopic("recoverer1");
+			return kafkaTemplate;
+		}
+
+		@Bean(name = KafkaStreamsDefaultConfiguration.DEFAULT_STREAMS_CONFIG_BEAN_NAME)
+		public KafkaStreamsConfiguration kStreamsConfigs(ApplicationContext context) {
+			Map<String, Object> props = new HashMap<>();
+			props.put(StreamsConfig.APPLICATION_ID_CONFIG, "testStreams");
+			props.put(StreamsConfig.BOOTSTRAP_SERVERS_CONFIG, this.brokerAddresses);
+			props.put(StreamsConfig.DEFAULT_KEY_SERDE_CLASS_CONFIG, Serdes.ByteArraySerde.class);
+			props.put(StreamsConfig.DEFAULT_VALUE_SERDE_CLASS_CONFIG, FailSerde.class);
+			props.put(StreamsConfig.DEFAULT_TIMESTAMP_EXTRACTOR_CLASS_CONFIG,
+					WallclockTimestampExtractor.class.getName());
+			props.put(StreamsConfig.COMMIT_INTERVAL_MS_CONFIG, "100");
+			props.put(StreamsConfig.DEFAULT_DESERIALIZATION_EXCEPTION_HANDLER_CLASS_CONFIG,
+					RecoveringDeserializationExceptionHandler.class);
+			props.put(RecoveringDeserializationExceptionHandler.KSTREAM_DESERIALIZATION_CONTEXT_ID, context.getId());
+			props.put(RecoveringDeserializationExceptionHandler.KSTREAM_DESERIALIZATION_RECOVERER, "recoverer");
```

Comment:
And why we can't just store a `recoverer()` instance in the properties to configure streams?

Reply:
It doesn't work - see my comment in the issue.